### PR TITLE
[SLICK-421] Added a caching layer for "page boot data" using the Wordpress Transients API.

### PR DESCRIPTION
**Pull request by ericperez**
This should reduce the number of calls to `/page-boot-data` from "once per pageview" to "once for every page URL at most once every 15 minutes," meaning every call to the same page in a 15 minute window will re-use the existing (cached) copy that will get stored in-memory on the Wordpress server.

### CODE REVIEW

File: svn/trunk/SlickEngagement_Plugin.php [hunk 1]
```diff
@@ -293,135 +262,199 @@ public function removeSemicolons($value)
...
+            echo "('" . addslashes($filmstrip_str) . "','" . addslashes($dcm_str) . "');" . "\n";
+            echo "\n</script>\n";
+        }
+    }
+
+    //Returns a name for the "page boot data" transient 
+    private function getTransientName()
+    {
+        DEFINE('PAGE_BOOT_DATA_TRANSIENT_PREFIX', 'slick_page_boot_');
+        $normalized_url = $_SERVER['HTTP_HOST'] . $_SERVER['REQUEST_URI'];
```
Comment (vkimcm) [inline svn/trunk/SlickEngagement_Plugin.php:315]: We need to add protocol in the url.  I found some cases where it was returning different boot data when protocol was missing.

File: svn/trunk/SlickEngagement_Plugin.php [hunk 2]
```diff
@@ -38,7 +33,7 @@ protected function initOptions()
 
     public function getPluginDisplayName()
     {
-        return 'Slick Engagement';
+        return 'Slicktream Engagement';
```
Comment (ectogon) [inline svn/trunk/SlickEngagement_Plugin.php]: Should be Slickstream

File: svn/trunk/SlickEngagement_Plugin.php [hunk 3]
```diff
@@ -293,135 +262,199 @@ public function removeSemicolons($value)
...
+            $this->echoSlickstreamComment("ERROR: Site Code missing from Plugin Settings; Slickstream services are disabled");
+            return;
+        }
+
+        global $wp;
+        
+        $transient_name = $this->getTransientName(); //Name for WP Transient Cache API Key
+
+        // Check for existing data in transient cache. If none, then fetch data from server.
+        if (false === ($boot_data_obj = get_transient($transient_name))) {
```
Comment (vkimcm) [inline svn/trunk/SlickEngagement_Plugin.php]: what do you think of adding a query param, so we have option to not read from cache and fetch from server to update cached data to latest instantly?

File: svn/trunk/SlickEngagement_Plugin.php [hunk 4]
```diff
@@ -293,135 +262,199 @@ public function removeSemicolons($value)
...
+        return $jsBlock;
+    }
+
+    //TODO: Clean this up / migrate to SR functions
     public function addSlickPageHeader()
     {
         global $post;
 
-        $this->getPageBootData();
+        $this->echoSlickstreamComment("Page Generated at " . $this->getCurrentTimeStampByTimeZone('America/New York') . " EST");
```
Comment (vkimcm) [inline svn/trunk/SlickEngagement_Plugin.php]: I tried running this plugin change, and got this error.  I would suggest testing it before it is ready for code review.

Fatal error: Uncaught Exception: DateTimeZone::__construct(): Unknown or bad timezone (America/New York)

Comment (ectogon) [reply svn/trunk/SlickEngagement_Plugin.php]: I bet it should be "America/New_York"

File: svn/trunk/SlickEngagement_Plugin.php [hunk 5]
```diff
@@ -293,135 +262,204 @@ public function removeSemicolons($value)
...
+        } else {
+            return null;
+        }
+    }
+
+    private function echoSlickBootJs($boot_data_obj) {
+      echo <<<JSBLOCK
+        <script>
+        window.\$slickBoot = window.\$slickBoot || {};
+        window.\$slickBoot.d = ' . json_encode($boot_data_obj) . ';
```
Comment (vkimcm) [inline svn/trunk/SlickEngagement_Plugin.php]: Fatal error: Uncaught Error: Object of class stdClass could not be converted to string
in /nas/content/live/slicksampleprd/wp-content/plugins/slick-engagement/SlickEngagement_Plugin.php on line 287

File: svn/trunk/SlickEngagement_Plugin.php [hunk 6]
```diff
@@ -293,135 +262,211 @@ public function removeSemicolons($value)
...
+            return;
+        }
+
+        global $wp;
+        
+        $transient_name = $this->getTransientName(); //Name for WP Transient Cache API Key
+
+        // If `slick-boot=1` is passed as a query param, force a re-fetch of the boot data from the server
+        $force_update_param_name = 'slick-boot';
+        $force_update_param = $_GET[$force_update_param_name];
```
Comment (vkimcm) [inline svn/trunk/SlickEngagement_Plugin.php]: this line is throwing warning when 'slick-boot' isn't set, so can we remove $force_update_param and check for it directly?
isset($_GET[$force_update_param_name])

Comment (ericperez) [reply svn/trunk/SlickEngagement_Plugin.php]: Yes. I have additional changes I'm going to commit today.

**Review by vkimcm — APPROVED** [review]

**Pull request merged**